### PR TITLE
WT-9720 Clean up and reorganize the fast-truncate code

### PR DESCRIPTION
**Pull request by sauclovian-wt**
Clean up and reorg the fast-truncate code.

Also, while passing by:
 - make two minor clarifications in test_checkpoint06
 - fix the test numbering in test_rts34

### CODE REVIEW

**Comment by sauclovian-wt** [comment]
I'm marking this as a draft for now; it is pretty much ready to review (definitely ready for initial review), but not quite ready to merge.

- I believe all the code changes are in place; unless further stress testing turns up more problems there won't be any more.
- I should run more stress testing before I call it ready to merge.
- I am planning to do a big audit of all the truncate-related comments I can find (particularly the ones that _aren't_ already in the changeset) tomorrow.
- However, I wanted to get this filed while it's still Friday afternoon in Sydney, in case anyone who has time for a quick look finds something I can fix/adjust tomorrow.
- I'm taking next week off, so on the one hand there's no rush but also if there's anything I can get sorted tomorrow so much the better.

There's also an architecture guide page in the works, which I will add here or file separately depending on how things progress.

**Comment by sauclovian-wt** [comment]
Oh, another note: this was originally intended to be cleanup/reorg only but there is at least one set of fixes (some truncate-related visibility checks were not handling prepared truncates correctly; this is what the new tests are about) and most likely the fix for WT-9713 is in here as well. (I saw the same problem yesterday and fixed it, but thought it was something I'd caused with the reorg.)

**Comment by sauclovian-wt** [comment]
Reviewing the comments turned up a problem, or perhaps two problems: I/we had overlooked that transaction abort discards ref->page_del while reasoning about the safety of doing so after reconciliation and while checking page_del in the case where internal page reconciliation needs to build a reference to the on-disk version of a just-instantiated page. In the first case there's a very small window when abort and reconciliation can both free the same page_del, which is quite bad, and in the second it's possible that abort can zap page_del while rec_child_deleted is inspecting it, which is still bad but not as bad.

I'm pretty sure both these conditions have been in develop for a good while, probably since WT-9252 was merged, and have escaped detection in testing so far.

I'm about to push fixes, but note that they haven't been heavily tested yet.

**Comment by agorrod** [comment]
@sauclovian-wt - the tests here passed. There was one failure, but it matched the signature of WT-9751. Is this ready for final review and (hopefully) merge?

**Comment by sauclovian-wt** [comment]
Yeah, I think so. I should probably run stress on it more locally still but there's no real reason to hold review for that.

File: src/include/btmem.h
Comment (agorrod) [inline src/include/btmem.h:1081]: Yay!!!

**Review by agorrod — APPROVED** [review]
I felt like I should have found some things to comment on, but it looks good to me.
I've got one celebratory remark, otherwise I'm happy with this change. I think it's a nice simplification.

**Comment by agorrod** [comment]
@quchenhao would you mind being a reviewer on this change? It's something David created, after we explored ways to improve fast truncate. There are some bug fixes, but also quite a lot of restructuring to make things more obvious (still not obvious, since truncate is complex).

**Comment by quchenhao** [comment]
@agorrod I'll have a look.

File: src/evict/evict_page.c
Comment (quchenhao) [inline src/evict/evict_page.c:522]: read barrier?

File: test/suite/test_truncate16.py
Comment (quchenhao) [inline test/suite/test_truncate16.py:50]: Why this is all disabled in all three tests?

File: src/btree/bt_delete.c
Comment (quchenhao) [inline src/btree/bt_delete.c:400]: stable is misleading here. I think this means globally visible?

File: src/btree/bt_delete.c
Comment (quchenhao) [inline src/btree/bt_delete.c:451]: A row that _was_ visible would have forced its page to be slow-truncated rather than fast-truncated.

I don't understand this. Does it mean we should slow truncate a page if any row is visible to the truncate operation?

File: src/btree/bt_delete.c
Comment (quchenhao) [inline src/btree/bt_delete.c:650]: Add a tombstone to row store doesn't mark the page dirty? I think we go through row_modify to do that and that should mark the page dirty too.

File: src/btree/bt_read.c
Comment (quchenhao) [inline src/btree/bt_read.c:168]: Should we update the oldest id to make it more likely that the page_del is globally visible?

File: src/btree/bt_read.c
Comment (quchenhao) [inline src/btree/bt_read.c:211]: I would change this to:

WT_ASSERT(previous_state != WT_REF_DISK || session, ref->page_del == NULL);
WT_ASSERT(previous_state != WT_REF_DISK || addr.del_set == false);

Comment (quchenhao) [reply src/btree/bt_read.c:211]: To be consistent within the code base how we write asserts

File: src/evict/evict_page.c
Comment (quchenhao) [inline src/evict/evict_page.c:533]: Can you create a ticket to track this work?

File: src/evict/evict_page.c
Comment (quchenhao) [inline src/evict/evict_page.c:562]: I don't understand the comment here.

**Comment by quchenhao** [comment]
@sauclovian-wt @agorrod I think I understand the change from a general point of view but not each individual detail. Awesome work @sauclovian-wt.

File: src/btree/bt_delete.c
Comment (sauclovian-wt) [reply src/btree/bt_delete.c]: Indeed. Fixed in [5431108](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/5431108959e98cdf9b3aa1b0fb0aa532b7ab0db8)

File: src/btree/bt_delete.c
Comment (sauclovian-wt) [reply src/btree/bt_delete.c:452]: I'm pretty sure what it was supposed to mean is: a row _with a stop time_ where the value was visible to the thread doing the truncate (that is, the stop time wasn't visible) would have forced a slow-truncate. Let me try to rewrite it so it's clearer. See [5431108](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/5431108959e98cdf9b3aa1b0fb0aa532b7ab0db8)

File: src/btree/bt_delete.c
Comment (sauclovian-wt) [reply src/btree/bt_delete.c:651]: For row-store it adds the updates to the update lists by hand rather than going through row_modify. That lets it avoid marking the page dirty. (The new VLCS code uses col_modify because the data structures involved are a lot more complicated.) Clarified in [5431108](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/5431108959e98cdf9b3aa1b0fb0aa532b7ab0db8)

File: src/btree/bt_read.c
Comment (sauclovian-wt) [reply src/btree/bt_read.c:171]: Good idea. Hadn't thought of that. Added in [13f7c10](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/13f7c10a82cf5980908c3540c1766f4108cc81c1)

File: src/btree/bt_read.c
Comment (sauclovian-wt) [reply src/btree/bt_read.c]: Fixed in [50ff2aa](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/50ff2aa9435ce60b6fddf0cad8611f9602a83e86)

File: src/evict/evict_page.c
Comment (sauclovian-wt) [reply src/evict/evict_page.c]: Yeah, probably a good idea; it's WT-9780. Updated the comments in [9f4de66](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/9f4de669cdc9003b518d3a587f95176468058569)

File: src/evict/evict_page.c
Comment (sauclovian-wt) [reply src/evict/evict_page.c]: There's a WT_REF_UNLOCK macro. It's not used here. It's not obvious if that was intentional or an oversight. It's not the same as the code here. (WT_REF_UNLOCK updates the ref hist and uses WT_PUBLISH, this does neither.) I didn't want to change it arbitrarily, but did want to leave a note for someone to think about it later. I wrote WT-9780 so this is also in its scope.

**Review by quchenhao — APPROVED** [review]
lgtm

**Comment by sauclovian-wt** [comment]
There's a remaining problem that came up while I was writing the architecture doc page; the atomic discard of `page_del` in instantiated pages ([aba5659](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/aba565916b0b84cb6a5c94ada39f8d9daeae9e55)) is all very well for protecting the post-reconciliation discard from rollback, and it's safe against parent reconciliation (because those can only happen at the same time if we're checkpointing the parent and evicting the child, and in that case eviction has the child ref locked) but there's a possible problem with prepare and commit. These fetch page_del and update it if it's not null, which is fine except that nothing guarantees that it finishes before the child reconciliation calls free. I think that post-reconciliation discard is going to need to lock the ref after all (if checkpointing rather than evicting). I'm going to sleep on it. Don't merge just yet...

**Comment by sauclovian-wt** [comment]
I think the test failure (or at least the one Alex forwarded me) is another manifestation of the same problem, except racing against the check in evict_page_clean_update. I think there's no getting around locking the ref there. This annoys me, but it's not really a problem (it's not by any means a common path...) I'll push that code.

**Comment by sueloverso** [comment]
While the content has been approved by @quchenhao and @agorrod, it is fairly out of date with develop and should be pulled forward. I tried to run against this branch with the mirror failure in WT-9772 based on @quchenhao's comment to see if I could repro but this branch is quite out of date with develop and the mirror config was not understood.

**Comment by sauclovian-wt** [comment]
Oops, I meant to do that and apparently forgot.

**Comment by agorrod** [comment]
There were three failures in stress testing, they were `hs_delete_reinsert_from_pos`, which I believe is WT-9763, which was merged last night. I think this is good to go.

Yay!

**Pull request merged**
